superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+- Disk storage, around ~65 GB per L2 chain of OP-Mainnet scale and history-size,
+  for an append-only log of block-hashes and event-log hashes per chain.
+
+Note that there does not have to be any indexing of hashes back to metadata,
+as the executing-messages include the block-number and event-log index, for fast lookups.
+And a binary-search over the append-only log can find the data quickly:
+there is no need for a more free key-value database.
+Optionally the hashes may be truncated to ~20 bytes, like ethereum addresses,
+to save on storage while still being reasonably cryptographically secure.
+
+
+## Responsibilities
+
+The responsibilities of the superchain backend are to:
+- (pre-)fetch executing messages of each chain.
+- Track cross-unsafe-head of each chain.
+- Track unsafe head of each chain, to handle reorgs.
+- Resolve block and message safety.
+- And possible future extensions, post-launch:
+  - Track shreds of each chain (inputs to intra-block message safety).
+  - Resolve block safety incl. cycles (intra-block message safety).
+
+## Implementation ideas
+
+### Verify as hashes, not messages
+
+Log-events can be huge, their size is variable.
+Importantly, an executing message already encodes what is being executed.
+Verifying should not require retrieval of the full initiating message.
+
+And there is no "partially valid":
+the log-event contract-origin address and log-event topics and data can be hashed into one commitment.
+
+Since initiating messages and executing messages are both log events,
+and assuming intra-block messaging is allowed by the protocol, we can:
+
+- use an executing message as initiating message
+- execute the same message
+- have a loop of executing messages between chains
+
+We should version the hash, and type it, such that we can:
+
+- Distinguish between "any" log event, and a log event that is an executing message and thus depends on another.
+- Persist the hashes, and still be able to tell them apart after changing the interop system.
+
+The hash should also be constructed like `H(origin,H(payload))` to not require the full payload,
+but just the payload hash, to construct the final message hash.
+
+### Data loading
+
+Transitive dependencies make a lookup unbounded.
+If a lookup will take long, we want to abort.
+A tx can always be retried, it should not lock up resources.
+Loading of initiating messages can take quite a while, this should not be the default critical path.
+
+Ideally we prefetch the data, and persist it efficiently for later lookups and validation.
+
+We can entirely separate all the data loading from validation:
+
+- **Loading**: Load log-events for each new block, transform into interop-hashes,
+  and append the block of hashes to a append-log.
+- **Validating**: Traverse the append-log, and for each executing hash, check for matching initiating hashes.
+  If at any point invalid, then truncate the log to remove any data that depends on it.
```

Comment:
I'm assuming we also truncate the log if there's a reorg caused by something other than logs not being valid (ie safe head update forces a reorg or there's an L1 reorg).